mopro-ffi/src/circom/mod.rs

```diff
@@ -29,6 +29,75 @@ use color_eyre::Result;
 
 use num_bigint::{BigInt, BigUint};
 
+pub type WtnsFn = fn(HashMap<String, Vec<num_bigint::BigInt>>) -> Vec<num_bigint::BigInt>;
+
+#[macro_export]
+macro_rules! circom_app {
+    () => {
+        static CIRCOM_CIRCUITS: once_cell::sync::Lazy<
+            std::collections::HashMap<String, mopro_ffi::WtnsFn>,
+        > = once_cell::sync::Lazy::new(|| set_circom_circuits());
+
+        fn generate_circom_proof(
+            in0: String,
+            in1: std::collections::HashMap<String, Vec<String>>,
+        ) -> Result<mopro_ffi::GenerateProofResult, mopro_ffi::MoproError> {
+            let name = std::path::Path::new(in0.as_str()).file_name().unwrap();
+            if let Some(witness_fn) = CIRCOM_CIRCUITS.get(name.to_str().unwrap()) {
+                mopro_ffi::generate_circom_proof_wtns(in0, in1, witness_fn.clone())
+            } else {
+                Err(mopro_ffi::MoproError::CircomError(
+                    format!(
+                        "Unknown ZKEY: {}. Have keys: {:?}",
+                        in0,
+                        CIRCOM_CIRCUITS.keys()
+                    )
+                    .to_string(),
+                ))
+            }
+        }
+
+        fn verify_circom_proof(
+            in0: String,
+            in1: Vec<u8>,
+            in2: Vec<u8>,
+        ) -> Result<bool, mopro_ffi::MoproError> {
+            mopro_ffi::verify_circom_proof(in0, in1, in2)
+        }
+
+        fn to_ethereum_proof(in0: Vec<u8>) -> mopro_ffi::ProofCalldata {
+            mopro_ffi::to_ethereum_proof(in0)
+        }
+
+        fn to_ethereum_inputs(in0: Vec<u8>) -> Vec<String> {
+            mopro_ffi::to_ethereum_inputs(in0)
+        }
+    };
+}
+
+/// Set the circuits that can be proven by the mopro library
+/// Provide the circuits that you want to be able to generate proofs for
+/// as a list of pairs of the form `zkey`, `wtns_fn`
+/// Where `zkey` is the name of the zkey file
+/// and `wtns_fn` is the function that generates the witness for the circuit
+/// NOTE: YOU CAN ONLY USE THIS MACROS ONCE IN YOUR CODEBASE, IN THE SAME MODULE AS `app!()`
+/// NOTE: TO USE THIS MACRO, YOU MUST HAVE THE `mopro-ffi/circom` FEATURE ENABLED
+#[macro_export]
+macro_rules! set_circom_circuits {
```

Comment:
Ok so we're using `set_circom_circuits!` to generate a function in the same scope as the above macro. The once-cell above invokes this function to set it's value.

I think we should remove this macro and the once-cell. Users should author the mapping function in rust like before.

Using macros makes it less clear _what_ is happening behind the scenes. It also introduces a non-standard syntax. e.g. even if i know rust i have to figure out this macro.

Reply:
Personally (and this is definitely a biased opinion) I though that this would be much more consistent to ask the users to invoke two macros, as opposed to asking to invoking a macro and then implementing a function, which at first does not seem to have clear relevance. 

Especially as rust does not have function overloads, I thought macros was a standard approach. 

As for complexity - macro is a simple solution, however, indeed, it only covers for a function and can be replaced by a function if the user wants.

My take is that if we want a user just use our library and simplify it as much as possible, macro abstracts all of the complexity, while allowing to also take a more complex approach.

I also made sure to add docs to macros (as well as will improve them to mention that a user can do it themselves).

Reply:
I will get rid of the `one-cell` as well as the static variable, this was made redundant by the split macro approach.

Reply:
I have significantly changed the underlying of the macros, so that we no longer use `one-cell` and instead they generate a match style function, similar to what was originally.

This effectively provides two ways for action for users:
They can simply use a macro (ideal for those who do not know how rust works)
```rust

mopro_ffi::set_circom_circuits! {
    ("multiplier2_final.zkey", multiplier2_witness),
    ("multiplier2_bls_final.zkey", multiplier2bls_witness),
    ("keccak256_256_test_final.zkey", keccak256256test_witness),
}
```rust
fn get_circom_wtns_fn(circuit: &str) -> Result<mopro_ffi::WtnsFn, mopro_ffi::MoproError> {
   match circuit {
      "circuit1.zkey" => Ok(circuit1_witness_fn),
     _ => Err(mopro_ffi::MoproError::CircomError(format!("Unknown ZKEY: {}", circuit).to_string()))
  }
}
```

The doc comment provides a guide to these both options, though I personally think that implementing a function is more complex than using a macro, especially as we already use a macro to initiate `mopro_ffi`.

Furthermore, the function requires error handling, arguably out of scope for the users and only adds future complexity.

Reply:
I like the change in the rust macro. Providing options as `(str, fn)` tuples is much more clear than just a list.

Having both options is probably the best solution 👍🏻